src/uct/tcp/tcp_ep.c

```diff
@@ -668,22 +669,26 @@ ucs_status_t uct_tcp_ep_handle_dropped_connect(uct_tcp_ep_t *ep,
     char str_local_addr[UCS_SOCKADDR_STRING_LEN];
     char str_remote_addr[UCS_SOCKADDR_STRING_LEN];
 
-    if ((io_status == UCS_ERR_NOT_CONNECTED) &&
-        ((ep->conn_state == UCT_TCP_EP_CONN_STATE_WAITING_ACK) ||
-         (ep->conn_state == UCT_TCP_EP_CONN_STATE_CONNECTING)) &&
-        ucs_test_all_flags(ep->ctx_caps, UCS_BIT(UCT_TCP_EP_CTX_TYPE_TX))) {
-        /* connection was closed by the peer during simultaneous connection
-         * resolution, this EP is the "loser" of the connection establishment.
-         * So, its socket fd has to be closed and fd from accept() will be used */
-        ucs_debug("tcp_ep %p: [%s <-> %s] connection was closed by the peer during "
-                  "simultaneous connection resolution, fd %d was closed", ep,
-                  ucs_sockaddr_str((const struct sockaddr*)&iface->config.ifaddr,
-                                   str_local_addr, UCS_SOCKADDR_STRING_LEN),
-                  ucs_sockaddr_str((const struct sockaddr*)&ep->peer_addr,
-                                   str_remote_addr, UCS_SOCKADDR_STRING_LEN), ep->fd);
-        uct_tcp_ep_mod_events(ep, 0, ep->events);
-        ucs_close_fd(&ep->fd);
-        return UCS_OK;
+    if (io_status == UCS_ERR_NOT_CONNECTED) {
+        if (((ep->conn_state == UCT_TCP_EP_CONN_STATE_WAITING_ACK) ||
```

Comment:
maybe revert the condition and `return io_status`, then no need for big `else` branch